opm/autodiff/BlackoilAquiferModel_impl.hpp

```diff
@@ -0,0 +1,138 @@
+    }
+
+
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>:: updateConnectionIntensiveQuantities() const
+    {
+        ElementContext elemCtx(ebosSimulator_);
+        const auto& gridView = ebosSimulator_.gridView();
+        const auto& elemEndIt = gridView.template end</*codim=*/0, Dune::Interior_Partition>();
+        for (auto elemIt = gridView.template begin</*codim=*/0, Dune::Interior_Partition>();
+             elemIt != elemEndIt;
+             ++elemIt)
+        {
+            elemCtx.updatePrimaryStencil(*elemIt);
+            elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+        }
+    }
+
+
+    // Protected function: Return number of aquifers in the model.
+    template<typename TypeTag>
+    int
+    BlackoilAquiferModel<TypeTag>:: numAquifers() const
+    {
+        return aquifers_.size();
+    }
+
+
+    // Protected function which calls the individual aquifer models
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>:: assembleAquiferEq(const SimulatorTimerInterface& timer)
+    {
+        for (auto aquifer = aquifers_.begin(); aquifer != aquifers_.end(); ++aquifer)
+        {
+            aquifer->assembleAquiferEq(ebosSimulator_, timer);
+        }
+    }
+
+    // Protected function
+    // some preparation work, mostly related to group control and RESV,
+    // at the beginning of each time step (Not report step)
+    template<typename TypeTag>
+    void BlackoilAquiferModel<TypeTag>:: prepareTimeStep(const SimulatorTimerInterface& timer)
+    {
+        // Here we can ask each carter tracy aquifers to get the current previous time step's pressure
+        for (auto aquifer = aquifers_.begin(); aquifer != aquifers_.end(); ++aquifer)
+        {
+            aquifer->before_time_step(ebosSimulator_, timer);
+        }
+    }
+
+    // Protected function: Returns a reference to the aquifers members in the model
+    template<typename TypeTag>
+    const std::vector< AquiferCarterTracy<TypeTag> >&
+    BlackoilAquiferModel<TypeTag>:: aquifers()
+    {
+        return aquifers_;
+    }
+
+
+    // Initialize the aquifers in the deck
+    template<typename TypeTag>
+    void
+    BlackoilAquiferModel<TypeTag>:: init(const Simulator& ebosSimulator, std::vector< AquiferCarterTracy<TypeTag> >& aquifers)
+    {
+        updateConnectionIntensiveQuantities();
+        const auto& deck = ebosSimulator.vanguard().deck();
+        const auto& eclState = ebosSimulator.vanguard().eclState();
+        
+        // Get all the carter tracy aquifer properties data and put it in aquifers vector
+        AquiferCT aquiferct = AquiferCT(eclState,deck);
+        Aquancon aquifer_connect = Aquancon(eclState.getInputGrid(), deck);
+
+        std::vector<AquiferCT::AQUCT_data> aquifersData = aquiferct.getAquifers();
+        std::vector<Aquancon::AquanconOutput> aquifer_connection = aquifer_connect.getAquOutput();
+
+        assert( aquifersData.size() == aquifer_connect.size() );

```

Comment:
@GitPaean Hi, it's strange that this line compiles in my system. But it is definitely not correct, as I intended it to be aquifer_connection instead of aquifer_connect. Thanks!

Reply:
> it's strange that this line compiles in my system.

you probably compiled it in release mode or with NDEBUG defined. asserts will be completely ignored in that case.

Reply:
> Hi, it's strange that this line compiles in my system.

If you only compile a Release build, you will not see any errors. This is because assert is a macro, that expands to nothing if NDEBUG is set, which is true for Release mode.

It is recommended to make a debug build as well, to trigger asserts when testing.

Reply:
Ah nice. Thanks for the information.